### PR TITLE
feat(apple): Automatic Instrumentation

### PR DESCRIPTION
**Pull request by philipphofmann**
Add docs for UI view controller and app start instrumtation, which
is still in preview.

### CODE REVIEW

File: src/platforms/apple/common/performance/instrumentation/automatic-instrumentation.mdx [hunk 1]
```diff
@@ -0,0 +1,82 @@
+---
+title: Automatic Instrumentation Preview
+sidebar_order: 10
+description: "Learn what transactions are captured after tracing is enabled."
+redirect_from:
+  - /performance/included-instrumentation/
```
Comment (PeloWriter) [inline src/platforms/apple/common/performance/instrumentation/automatic-instrumentation.mdx:6]: I don't believe we need a redirect for this page?

File: src/platforms/apple/common/performance/instrumentation/automatic-instrumentation.mdx [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+---
+
+<Note>
+
+This is a preview API, available in `7.2.0-beta.0` and above. Features available in a preview API are still in-progress and may have bugs. We recognize the irony. The API may be renamed, changed, or removed in a future version.
```
Comment (PeloWriter) [inline src/platforms/apple/common/performance/instrumentation/automatic-instrumentation.mdx:11]:
```suggestion
This is a preview API, available in `7.2.0-beta.0` and above. Features available in a preview API are still in progress and may have bugs. We recognize the irony. The API may be renamed, changed, or removed in a future version.
```

File: src/platforms/apple/common/performance/instrumentation/automatic-instrumentation.mdx [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+Capturing transactions requires that you first <PlatformLink to="/performance/">set up performance monitoring</PlatformLink> if you haven't already.
+
+</Note>
+
+## UIViewController Instrumentation
+
+This feature is available for iOS, tvOS, and Mac Catalyst and only works for UIViewControllers.
```
Comment (PeloWriter) [inline src/platforms/apple/common/performance/instrumentation/automatic-instrumentation.mdx:19]:
```suggestion
This feature is available for iOS, tvOS, and Mac Catalyst, and only works for UIViewControllers.
```

File: src/platforms/apple/common/performance/instrumentation/automatic-instrumentation.mdx [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+The UIViewController Instrumentation, once enabled, captures transactions whenever your app loads a ViewController. The SDK sets the transaction name to the name of the ViewController, including the module, for example, `Your_App.MainViewController`, and the transaction operation to `ui.rendering`. The SDK creates spans to give you insight into how long each of the following methods took.
```
Comment (PeloWriter) [inline src/platforms/apple/common/performance/instrumentation/automatic-instrumentation.mdx:21]:
```suggestion
The UIViewController Instrumentation, once enabled, captures transactions whenever your app loads a ViewController. The SDK sets the transaction name to the name of the ViewController, including the module; for example, `Your_App.MainViewController`, and the transaction operation to `ui.rendering`. The SDK creates spans to provide insight into the length of time each consumed by each of the following methods.
```

File: src/platforms/apple/common/performance/instrumentation/automatic-instrumentation.mdx [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+![UIViewController Transaction](ui-view-controller-transaction.png)
+
+If you'd like to opt out of this feature, you can do so using options:
```
Comment (PeloWriter) [inline src/platforms/apple/common/performance/instrumentation/automatic-instrumentation.mdx:25]:
```suggestion
You can opt out of this feature using options:
```

File: src/platforms/apple/common/performance/instrumentation/automatic-instrumentation.mdx [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    options.dsn = @"___PUBLIC_DSN___";
+    options.enableAutoUIPerformanceTracking = NO;
+}];
+```
+
+## App Start Instrumentation
+
+This feature is available for iOS, tvOS, and Mac Catalyst.
+
+The App Start Instrumentation gives you insight into how long your application takes to launch. It adds spans for different phases of the application launch to the first auto-generated UI transaction.
```
Comment (PeloWriter) [inline src/platforms/apple/common/performance/instrumentation/automatic-instrumentation.mdx:49]:
```suggestion
The App Start Instrumentation provides insight into how long your application takes to launch. It adds spans for different phases, from the application launch to the first auto-generated UI transaction.
```

File: src/platforms/apple/common/performance/instrumentation/automatic-instrumentation.mdx [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+For this feature to work, ensure that `AutoUIPerformanceTracking` is enabled and set a <PlatformLink to="/configuration/sampling/#setting-a-uniform-sample-rate">`tracesSampleRate`</PlatformLink> or a <PlatformLink to="/configuration/sampling/#setting-a-sampling-function">`tracesSampler`</PlatformLink>. The SDK differentiates between a cold and a warm start and doesn't track hot starts/resumes.
```
Comment (PeloWriter) [inline src/platforms/apple/common/performance/instrumentation/automatic-instrumentation.mdx:50]:
```suggestion
To enable this feature enable `AutoUIPerformanceTracking` and set a <PlatformLink to="/configuration/sampling/#setting-a-uniform-sample-rate">`tracesSampleRate`</PlatformLink> or a <PlatformLink to="/configuration/sampling/#setting-a-sampling-function">`tracesSampler`</PlatformLink>. The SDK differentiates between a cold and a warm start, and doesn't track hot starts/resumes.
```

File: src/platforms/apple/common/performance/instrumentation/automatic-instrumentation.mdx [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+* __Cold start__: App launched for the first time, after a reboot or update. The app is not in memory, and no process exists.
```
Comment (PeloWriter) [inline src/platforms/apple/common/performance/instrumentation/automatic-instrumentation.mdx:51]:
```suggestion
* __Cold start__: App launched for the first time, after a reboot or update. The app is not in memory and no process exists.
```

File: src/platforms/apple/common/performance/instrumentation/automatic-instrumentation.mdx [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+* __Warm start__: App launched at least once, is partially in memory, and no process exists.
+
+The SDK uses the process start time as the beginning of the app start and the [`UIWindowDidBecomeVisibleNotification`][UIWindow] as the end. It creates the following spans:
+* __Pre Main__: From the process start time to the runtime init.
+* __UIKit and Application Init__: From the runtime init to the [`didFinishLaunchingNotification`][didFinishLaunching].
+* __Initial Frame Render__: From the [`didFinishLaunchingNotification`][didFinishLaunching] to [`UIWindowDidBecomeVisibleNotification`][UIWindow].
+
+![App Start Transaction](app-start-transaction.png)
+
+If you'd like to opt out of this feature, you can do so using options:
```
Comment (PeloWriter) [inline src/platforms/apple/common/performance/instrumentation/automatic-instrumentation.mdx:61]:
```suggestion
You can opt out of this feature using options:
```

**Review by PeloWriter — COMMENTED** [review]
I think we also need to modify the `common/performance/index` file at line 36. Can you take a look? LMK if you need help.

File: src/platforms/apple/common/performance/instrumentation/automatic-instrumentation.mdx (hunk 1 above)
Comment (philipphofmann) [reply src/platforms/apple/common/performance/instrumentation/automatic-instrumentation.mdx:6]: Ups a good catch. No, we don't need that. Was a copy-paste error.

**Comment by philipphofmann** [comment]
> I think we also need to modify the `common/performance/index` file at line 36. Can you take a look? LMK if you need help.

Thanks, @PeloWriter, for the review 😀. I don't get it to which file you are referring to. Can you please provide a link?

File: src/platforms/apple/common/performance/instrumentation/automatic-instrumentation.mdx [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+
+</Note>
+
+## UIViewController Instrumentation
+
+This feature is available for iOS, tvOS, and Mac Catalyst, and only works for UIViewControllers.
+
+The UIViewController Instrumentation, once enabled, captures transactions whenever your app loads a ViewController. The SDK sets the transaction name to the name of the ViewController, including the module; for example, `Your_App.MainViewController`, and the transaction operation to `ui.load`. The SDK creates spans to provide insight into the length of time each consumed by each of the following methods.
+
+![UIViewController Transaction](ui-view-controller-transaction.png)
```
Comment (philipphofmann) [inline src/platforms/apple/common/performance/instrumentation/automatic-instrumentation.mdx:21]: `h`:  Update these screenshots as the operation names changed after merging https://github.com/getsentry/sentry-cocoa/pull/1164

**Comment by marandaneto** [comment]
> > I think we also need to modify the `common/performance/index` file at line 36. Can you take a look? LMK if you need help.
> 
> Thanks, @PeloWriter, for the review 😀. I don't get it to which file you are referring to. Can you please provide a link?

probably https://github.com/getsentry/sentry-docs/blame/master/src/platforms/common/performance/index.mdx#L36

**Comment by PeloWriter** [comment]
Yes, @marandaneto has the right lines and file. Sorry, @philipphofmann - I'll be more clear in future. On that line we're explicitly stating that apple SDKs don't support automatic instrumentation, so we need to either remove or modify the content. Here's my suggested edit:

<PlatformSection supported={["apple"]}>

Automatic instrumentation for monitoring the performance of your application is currently in preview. You can also use <PlatformLink to="/performance/instrumentation/custom-instrumentation/">custom instrumentation to capture transactions</PlatformLink>.

</PlatformSection>

File: src/platforms/apple/common/performance/instrumentation/automatic-instrumentation.mdx [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+    options.enableAutoUIPerformanceTracking = NO;
+}];
+```
+
+## App Start Instrumentation
+
+This feature is available for iOS, tvOS, and Mac Catalyst.
+
+The App Start Instrumentation provides insight into how long your application takes to launch. It adds spans for different phases, from the application launch to the first auto-generated UI transaction.
+To enable this feature enable `AutoUIPerformanceTracking` and set a <PlatformLink to="/configuration/sampling/#setting-a-uniform-sample-rate">`tracesSampleRate`</PlatformLink> or a <PlatformLink to="/configuration/sampling/#setting-a-sampling-function">`tracesSampler`</PlatformLink>. The SDK differentiates between a cold and a warm start, and doesn't track hot starts/resumes.
```
Comment (marandaneto) [inline src/platforms/apple/common/performance/instrumentation/automatic-instrumentation.mdx:48]:
```suggestion
To enable this feature enable `AutoUIPerformanceTracking`. The SDK differentiates between a cold and a warm start, and doesn't track hot starts/resumes.
```

the note already mentions that you should set up the performance feature first

File: src/platforms/apple/common/performance/instrumentation/automatic-instrumentation.mdx [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+![App Start Transaction](app-start-transaction.png)
+
+You can opt out of this feature using options:
+
+```swift {tabTitle:Swift}
+import Sentry
+
+SentrySDK.start { options in
+    options.dsn = "___PUBLIC_DSN___"
+    options.enableAppStartMeasuring = false
```
Comment (marandaneto) [inline src/platforms/apple/common/performance/instrumentation/automatic-instrumentation.mdx:66]: I thought that we decided not offering such an option as it's bounded to the `enableAutoUIPerformanceTracking` anyway.

Comment (philipphofmann) [reply src/platforms/apple/common/performance/instrumentation/automatic-instrumentation.mdx:66]: Okay, then I'm going to remove it here and create an issue for removing it in Cocoa. We can always add such a feature later on.

**Review by marandaneto — APPROVED** [review]
LGTM

**Pull request merged**